presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -409,63 +409,69 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                     }
-                    else {
-                        joinFilterBuilder.add(conjunct);
+
+                    Symbol rightSymbol = symbolForExpression(rightExpression);
+                    if (!node.getRight().getOutputSymbols().contains(rightSymbol)) {
+                        rightProjections.put(rightSymbol, rightExpression);
                     }
-                }
 
-                Optional<Expression> newJoinFilter = Optional.of(combineConjuncts(joinFilterBuilder.build()));
-                if (newJoinFilter.get() == TRUE_LITERAL) {
-                    newJoinFilter = Optional.empty();
+                    equiJoinClauses.add(new JoinNode.EquiJoinClause(leftSymbol, rightSymbol));
                 }
-
-                if (node.getType() == INNER && newJoinFilter.isPresent() && equiJoinClauses.isEmpty()) {
-                    // if we do not have any equi conjunct we do not pushdown non-equality condition into
-                    // inner join, so we plan execution as nested-loops-join followed by filter instead
-                    // hash join.
-                    // todo: remove the code when we have support for filter function in nested loop join
-                    postJoinPredicate = combineConjuncts(postJoinPredicate, newJoinFilter.get());
-                    newJoinFilter = Optional.empty();
+                else {
+                    joinFilterBuilder.add(conjunct);
                 }
+            }
 
+            Optional<Expression> newJoinFilter = Optional.of(combineConjuncts(joinFilterBuilder.build()));
+            if (newJoinFilter.get() == TRUE_LITERAL) {
+                newJoinFilter = Optional.empty();
+            }
+
+            if (node.getType() == INNER && newJoinFilter.isPresent() && equiJoinClauses.isEmpty()) {
+                // if we do not have any equi conjunct we do not pushdown non-equality condition into
+                // inner join, so we plan execution as nested-loops-join followed by filter instead
+                // hash join.
+                // todo: remove the code when we have support for filter function in nested loop join
+                postJoinPredicate = combineConjuncts(postJoinPredicate, newJoinFilter.get());
+                newJoinFilter = Optional.empty();
+            }
+
+            boolean filtersEquivalent =
+                    newJoinFilter.isPresent() == node.getFilter().isPresent() &&
+                            (!newJoinFilter.isPresent() || expressionEquivalence.areExpressionsEquivalent(session, newJoinFilter.get(), node.getFilter().get(), types));
+
+            if (leftSource != node.getLeft() ||
+                    rightSource != node.getRight() ||
+                    !expressionEquivalence.areExpressionsEquivalent(session, newJoinPredicate, joinPredicate, types) ||
+                    !filtersEquivalent ||
+                    !ImmutableSet.copyOf(equiJoinClauses).equals(ImmutableSet.copyOf(node.getCriteria()))) {
```

Comment:
should not we return not rewritten node when this condition is not met?

Reply:
this if checks if anything related to join or children changed. If it did, then join node needs to rewritten